libraries/botbuilder-dialogs-adaptive/src/sequenceContext.ts

```diff
@@ -0,0 +1,203 @@
+
+            // Apply each queued set of changes
+            for (let i = 0; i < queue.length; i++) {
+
+                // Apply plan changes
+                const change = queue[i];
+                switch (change.changeType) {
+                    case StepChangeType.insertSteps:
+                    case StepChangeType.insertStepsBeforeTags:
+                    case StepChangeType.appendSteps:
+                        await this.updateSequence(change);
+                        break;
+                    case StepChangeType.endSequence:
+                        if (this.steps.length > 0) {
+                            this.steps.splice(0, this.steps.length);
+                            await this.emitEvent(AdaptiveEventNames.sequenceEnded, undefined, false);
+                        }
+                        break;
+                    case StepChangeType.replaceSequence:
+                        if (this.steps.length > 0) {
+                            this.steps.splice(0, this.steps.length);
+                        }
+                        await this.updateSequence(change);
+                        break; 
+                }
+            }
+
+            // Apply any new queued up changes
+            await this.applyChanges();
+            return true;
+        }
+
+        return false;
+    }
+
+    public insertSteps(steps: StepState[]): this {
+        this.queueChanges({ changeType: StepChangeType.insertSteps, steps: steps });
+        return this;
+    }
+
+    public insertStepsBeforeTags(tags: string[], steps: StepState[]): this {
+        this.queueChanges({ changeType: StepChangeType.insertStepsBeforeTags, steps: steps, tags: tags });
+        return this;
+    }
+
+    public appendSteps(steps: StepState[]): this {
+        this.queueChanges({ changeType: StepChangeType.appendSteps, steps: steps });
+        return this;
+    }
+
+    public endSequence(): this {
+        this.queueChanges({ changeType: StepChangeType.endSequence });
+        return this;
+    }
+
+    public replaceSequence(steps: StepState[]): this {
+        this.queueChanges({ changeType: StepChangeType.replaceSequence, steps: steps });
+        return this;
+    }
+
+    private async updateSequence(change: StepChangeList): Promise<void> {
+        // Initialize sequence if needed
+        const newSequence = this.steps.length == 0;
+
+        // Update sequence
+        switch (change.changeType) {
+            case StepChangeType.insertSteps:
+                Array.prototype.unshift.apply(this.steps, change.steps);
+                break;
+            case StepChangeType.insertStepsBeforeTags:
+                let inserted = false;
+                if (Array.isArray(change.tags)) {
+                    for (let i = 0; i < this.steps.length; i++) {
+                        if (this.stepHasTags(this.steps[i], change.tags)) {
+                            const args = ([i, 0] as any[]).concat(change.steps);
```

Comment:
I (and probably future developers) would appreciate some comments on the intention of lines 167 thru 170

Reply:
sure